extensions/scarb-doc/src/types.rs

```diff
@@ -0,0 +1,354 @@
+            .collect();
+
+        let module_enums = db.module_enums(module_id).unwrap();
+        let enums = module_enums
+            .iter()
+            .map(|(id, node)| Enum::new(db, *id, node))
+            .collect();
+
+        let module_type_aliases = db.module_type_aliases(module_id).unwrap();
+        let type_aliases = module_type_aliases
+            .iter()
+            .map(|(id, node)| TypeAlias::new(db, *id, node))
+            .collect();
+
+        let module_impl_aliases = db.module_impl_aliases(module_id).unwrap();
+        let impl_aliases = module_impl_aliases
+            .iter()
+            .map(|(id, node)| ImplAlias::new(db, *id, node))
+            .collect();
+
+        let module_traits = db.module_traits(module_id).unwrap();
+        let traits = module_traits
+            .iter()
+            .map(|(id, node)| Trait::new(db, *id, node))
+            .collect();
+
+        let module_impls = db.module_impls(module_id).unwrap();
+        let impls = module_impls
+            .iter()
+            .map(|(id, node)| Impl::new(db, *id, node))
+            .collect();
+
+        let module_extern_types = db.module_extern_types(module_id).unwrap();
+        let extern_types = module_extern_types
+            .iter()
+            .map(|(id, node)| ExternType::new(db, *id, node))
+            .collect();
+
+        let module_extern_functions = db.module_extern_functions(module_id).unwrap();
+        let extern_functions = module_extern_functions
+            .iter()
+            .map(|(id, node)| ExternFunction::new(db, *id, node))
+            .collect();
+
+        let module_submodules = db.module_submodules(module_id).unwrap();
+        let submodules = module_submodules
+            .iter()
+            .map(|(id, _node)| Self::new(db, ModuleId::Submodule(*id)))
+            .collect();
+
+        Self {
+            module_id,
+            full_path: module_id.full_path(db),
+            submodules,
+            constants,
+            uses,
+            free_functions,
+            structs,
+            enums,
+            type_aliases,
+            impl_aliases,
+            traits,
+            impls,
+            extern_types,
+            extern_functions,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ItemData {
+    pub name: String,
+    pub full_path: String,
+
+    pub doc: Option<String>,
+    pub definition: String,
+
+    pub text: String,
+}
+
+impl ItemData {
+    pub fn new(db: &dyn DefsGroup, id: ModuleItemId, node: &impl TypedSyntaxNode) -> Self {
+        Self {
+            name: id.name(db).into(),
+            full_path: id.full_path(db),
+            doc: db.get_item_documentation(LookupItemId::ModuleItem(id)),
+            definition: db.get_item_definition(LookupItemId::ModuleItem(id)),
+            text: node.as_syntax_node().get_text_without_trivia(db.upcast()),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Constant {
```

Comment:
I feel like all of those defs + impls are repetitive and could be done with a macro with like 3 input vars (pointer type, module type and the ModuleItemId)

Reply:
ikr, but they are planned to change and have some separate attributes, so I kept them like this for now